CRM/Dedupe/Merger.php

```diff
@@ -1631,41 +1631,48 @@ public static function getRowsElementsAndInfo(int $mainID, int $otherID, bool $c
-      $other['contact_sub_type'] ?? NULL,
-      $checkPermissions ? CRM_Core_Permission::EDIT : FALSE
-    );
-
-    foreach ($otherTree as $gid => $group) {
-      if (!isset($group['fields'])) {
-        continue;
-      }
-
-      foreach ($group['fields'] as $fid => $field) {
-        $mainContactValue = $mainTree[$gid]['fields'][$fid]['customValue'] ?? NULL;
-        $otherContactValue = $otherTree[$gid]['fields'][$fid]['customValue'] ?? NULL;
+    $filters = [
+      'extends' => $main['contact_type'],
+      'is_active' => TRUE,
+      'is_multiple' => FALSE,
+      'extends_entity_column_value' => NULL,
+    ];
+    if (!empty($other['contact_sub_type'])) {
+      $filters['extends_entity_column_value'] = array_merge([NULL], $other['contact_sub_type']);
+    }
+    $otherTree = CRM_Core_BAO_CustomGroup::getAll($filters, $checkPermissions ? CRM_Core_Permission::EDIT : NULL);
+
+    foreach ($otherTree as $group) {
+      $mainCustomValues = Contact::get(FALSE)
+        ->addWhere('id', '=', $mainID)
+        ->addSelect($group['name'] . '.*')
+        ->execute();
+      $otherCustomValues = Contact::get(FALSE)
+        ->addWhere('id', '=', $otherID)
+        ->addSelect($group['name'] . '.*')
+        ->execute();
+      foreach ($group['fields'] as $field) {
+        $fid = $field['id'];
+        $apiFieldName = "{$group['name']}.{$field['name']}";
         if (in_array($fid, $compareFields['custom'])) {
-          $rows["custom_group_$gid"]['title'] ??= $group['title'];
+          $rows["custom_group_{$group['id']}"]['title'] ??= $group['title'];
 
-          if ($mainContactValue) {
-            foreach ($mainContactValue as $valueId => $values) {
-              $rows["move_custom_$fid"]['main'] = CRM_Core_BAO_CustomField::displayValue($values['data'], $fid);
-            }
+          foreach ($mainCustomValues as $values) {
+            $customValue = $values[$apiFieldName] ?? NULL;
+            $rows["move_custom_$fid"]['main'] = CRM_Core_BAO_CustomField::displayValue($customValue, $fid);
           }
           $value = 'null';
-          if ($otherContactValue) {
-            foreach ($otherContactValue as $valueId => $values) {
-              $rows["move_custom_$fid"]['other'] = CRM_Core_BAO_CustomField::displayValue($values['data'], $fid);
-              if ($values['data'] === 0 || $values['data'] === '0') {
-                $values['data'] = $qfZeroBug;
-              }
-              $value = ($values['data']) ? $values['data'] : $value;
+          foreach ($otherCustomValues as $values) {
+            $customValue = $values[$apiFieldName] ?? NULL;
+            $rows["move_custom_$fid"]['other'] = CRM_Core_BAO_CustomField::displayValue($customValue, $fid);
+            if ($customValue === 0 || $customValue === '0' || $customValue === FALSE) {
+              $customValue = $qfZeroBug;
             }
+            // FIXME: Not changed during refactor but this looks wrong: should be `??` not `?:`
```

Comment:
I don't think this comment is quite right but I think it's easier to put up a follow up PR than to explain in a comment